docs/rfcs/20220207_connections_per_ip.md

```diff
@@ -0,0 +1,160 @@
+number of unique clients maybe high (so a flat store is risky), but the entries
+are small (so a std::map would generate a lot of tiny allocations).
+
+### Optimization: caching tokens
+
+In the case where the client is not close to exceeding its limits, we can avoid
+the overhead of a cross-core call by caching some per-client tokens on each
+shard.
+
+This cache can be pre-primed on the first client connection, if the total
+tokens for the client is greater than the core count: that way, subsequent
+connections from the same client will usually be accepted without the need for
+a cross-core call.
+
+### Handling a configuration change
+
+Changes to the configuration properties affect all outstanding
+client_conn_quota objects: the `total` attribute must be recalculated.
+
+If `total` is decreased, then it is also necessary to call out to all other
+shards and revoke any cached tokens that are held in excess of the new limit.
+
+## Impact
+
+When this feature is enabled, a per-connection latency overhead is added, as
+the shard handling the connection must dispatch a call to the shard that owns
+the client IP state.
+
+On a system where the connection count limit is significantly larger than the
+core count, this overhead will be avoided in most cases, as each core will
+store some locally cached tokens to allow them to accept connections.
+
+
+# Guide-level explanation
+
+## How do we teach this?
+
+* Presume reader knows what a TCP connection is
+* Explain how Kafka clients map to TCP connections: that a producer or consumer
+  usually opens a single connection and uses it over a long period of time, but
+a program that constructs many Kafka clients in parallel might 
+* Explain why it's bad to have an unbounded number of connections from a
+  client:
+  * Each connection has some memory overhead that might exhaust redpanda memory
+  * Each connection counts against OS-level limits (`ulimit`)
+  * It's probably not what the user intended for efficient operation, as the
+    Kafka protocol generally re-uses a smaller number of connections.
+* Introduce the new configuration properties, including the syntax of the
+  `overrides` property
+  * Mention the caveat that decreasing the limits does not terminate any
+    currently open connections.
+
+## Interaction with other features
+
+* Dependency on centralized config for live updates
+
+## Drawbacks
+
+### Latency overhead on first message in a new connection:
+
+Mitigating factors:
+* Sensible workloads generally operate on established connections
+* No overhead if feature is not enabled.
+
+## Alternatives Considered
+
+### Do nothing
+
+A client may exhaust server resources by opening a very large number of
+connections.  This is okay as long as all clients are well-behaved and the
+system is carefully sized for its workload, neither of which are reliably true
+in real life deployments.
+
+### Total connection count, rather than per-client
```

Comment:
I think it'd be beneficial to rephrase per-client in this document to per-server. Client may mistakenly lead people to believe that they will not be impacted if they use 30 clients on one server from many different applications, whereas per-server more clearly identifies that as a risk they will need to consider (it may be worth calling out this risk, too).

Reply:
You're right, I'm abusing "client" to mean several things in this doc.